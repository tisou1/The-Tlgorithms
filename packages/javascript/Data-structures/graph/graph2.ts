


export class Graph2<T = unknown> {
  noOfVertices: T[]
  AdjList: Map<T, T[]>

  constructor(noOfVertices) {
    this.noOfVertices = noOfVertices
    this.AdjList = new Map()
  }

  //添加顶点
  addVertex(v: T) {
    this.AdjList.set(v, [])
  }
  //添加线
  addEdge(v: T, w: T) {
    this.AdjList.get(v)?.push(w)

    this.AdjList.get(w)?.push(v)
  }

  printGraph(output = (value: string) => console.log(value)){
    const getKeys = this.AdjList.keys()

    for(const i of getKeys) {
      const getValue = this.AdjList.get(i)
      let conc = ''

      for(const j of getValue as T[]) {
        conc += j + ' '
      }

      output(i + ' -> ' + conc)
    }
  }
}