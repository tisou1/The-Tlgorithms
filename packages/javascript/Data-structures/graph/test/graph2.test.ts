import { vi, expect, it, describe, beforeEach } from 'vitest'
import { Graph2 } from '../graph2'

describe('Test graph2', () => {
  const vertices = ['A', 'B', 'C', 'D', 'E', 'F']
  const graph = new Graph2(vertices.length)

  // adding vertices
  for (let i = 0; i < vertices.length; i++) {
    graph.addVertex(vertices[i])
  }

    // adding edges
    graph.addEdge('A', 'B')
    graph.addEdge('A', 'D')
    graph.addEdge('A', 'E')
    graph.addEdge('B', 'C')
    graph.addEdge('D', 'E')
    graph.addEdge('E', 'F')
    graph.addEdge('E', 'C')
    graph.addEdge('C', 'F')
  
    
  it('check one', () => {
    const mockFn = vi.fn()
    graph.printGraph(mockFn)

    //函数被调用的次数
    expect(mockFn.mock.calls.length).toBe(vertices.length)
    //mock.calls是包含每个调用的所有参数的数组
    const adjListArr = mockFn.mock.calls.map( v => v[0])
    expect(adjListArr).toMatchSnapshot()

    expect(adjListArr).toEqual([
      'A -> B D E ',
      'B -> A C ',
      'C -> B E F ',
      'D -> A E ',
      'E -> A D F C ',
      'F -> E C '
    ])
  })
})