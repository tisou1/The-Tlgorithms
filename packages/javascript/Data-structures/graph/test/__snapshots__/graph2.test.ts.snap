// Vitest Snapshot v1

exports[`Test graph2 > check one 1`] = `
[
  "A -> B D E ",
  "B -> A C ",
  "C -> B E F ",
  "D -> A E ",
  "E -> A D F C ",
  "F -> E C ",
]
`;
